hadoop-yarn-project/hadoop-yarn/bin/FederationStateStore/SQLServer/FederationStateStoreStoreProcs.sql

```diff
@@ -111,12 +111,37 @@ IF OBJECT_ID ( '[sp_getApplicationsHomeSubCluster]', 'P' ) IS NOT NULL
 GO
 
 CREATE PROCEDURE [dbo].[sp_getApplicationsHomeSubCluster]
+    @limit int,
+    @homeSubCluster VARCHAR(256)
 AS BEGIN
     DECLARE @errorMessage nvarchar(4000)
 
     BEGIN TRY
-        SELECT [applicationId], [homeSubCluster], [createTime]
-        FROM [dbo].[applicationsHomeSubCluster]
+        IF @homeSubCluster = ''
+            SELECT
+                [applicationId],
+                [homeSubCluster],
+                [createTime]
+            FROM(SELECT
+                     [applicationId],
+                     [homeSubCluster],
+                     [createTime],
+                     row_number() over(order by [createTime] desc) AS app_rank
+                 FROM [dbo].[applicationsHomeSubCluster]) AS t
+            WHERE app_rank <= @limit;
+        ELSE
+            SELECT
+                [applicationId],
+                [homeSubCluster],
+                [createTime]
+            FROM(SELECT
+                     [applicationId],
+                     [homeSubCluster],
+                     [createTime],
+                     row_number() over(partition by [homeSubCluster] order by [createTime] desc) AS app_rank
+                 FROM [dbo].[applicationsHomeSubCluster]
+                 WHERE [homeSubCluster] = @homeSubCluster) AS t
```

Comment:
Why not 
```
WHERE [homeSubCluster] = @homeSubCluster OR @homeSubCluster = ''
```

Reply:
My code is a bit complicated, your approach is really good.